text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,414 @@
+
+Once you found what you're looking for, having a link to an item's documentation from the source code allows to add the missing connection between the source code viewer and the doc pages.
+
+Another important note is that if documentation is split in multiple places or requires an external tool for the source code navigation, it's actually slowing down our users quite a lot.
+
+A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.
+
+Here is a video showing the feature in action:
+
+https://user-images.githubusercontent.com/3050060/114622354-21307b00-9cae-11eb-834d-f6d8178a37bd.mp4
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature **only** targets the source code pages.
+
+It will be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.
+
+## Link generation
+
+There are two kinds of links:
+
+ * Jump to documentation.
+ * Jump to definition.
+
+### Jump to documentation
+
+All items with a documentation page will have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks won't (explications below).
+
+The link will be generated on their ident/name where they are declared, not when they are in use. For example:
+
+```rust
+pub struct Foo;
+
+let x = Foo;
+```
+
+On `Foo` definition (`pub struct Foo`), we will generate a link to its documentation page whereas on its usage (`x = Foo`), we will generate a link to jump to its definition.
+
+This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we will generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.
+
+For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.
+
+### Jump to definition
+
+A small preamble before listing the items we should generate links for: if an item which will get a link is preceded by a path, each part of the path will be a link to its target. For example:
```

Comment:
```suggestion
We will generate link for each path segment of a path to its respective definition. For example:
```

The phrasing of this paragraph confuses me a bit. Minor: I don't understand what “preamble” means in this context.
More significantly, however: You write ”*an item […] preceded by a path*” and – from what I infer – call the `D` in the example below an *item* and the `b::c` a *path*.
The quote is not correct, the whole snippet `b::c::D` is a path and `b`, `c` & `D` are all path *segments*. An item cannot be “*preceded by a path*”. `D` is not an item, it is a path segment referring to an item.
I don't want to sound overly pedantic.

Obviously my new phrasing overlaps a bit with the subsection **Jump to definition**/**Paths**. So one or the other can be cut down in size, I'd say.

Reply:
Preventing confusion is a very good thing, good catch!